theano/gof/utils.py

```diff
@@ -217,14 +217,21 @@ def __call__(cls, *args, **kwargs):
         props = cls.__dict__.get('__props__', None)
         if props is None:
             return super(MetaObject, cls).__call__(*args, **kwargs)
+
         key = [cls]
         for arg in args:
             arg = make_hashable(arg)
             key.append(arg)
+            arg_dtype = get_arg_dtype(arg)
+            if arg_dtype:
+                key.extend(arg_dtype)
         for k, v in kwargs.items():
             key.append(k)
             v = make_hashable(v)
             key.append(v)
+            v_dtype = get_arg_dtype(v)
+            if v_dtype:
+                key.extend(v_dtype)
```

Comment:
I did not fully understand what you mean here.  I still have the value in the key for caching. The reason that I added the ```dtype``` was explained in my yesterday's comment. We had an error where an object gave outputs with different ```dtypes``` in two cases.``` dtype``` was added to make sure the same object is not used in these two cases. 

Yes I will test both cases.